src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -444,5 +431,76 @@ private static string GetDirectoryName(string storeName)
 
             return storeName.ToLowerInvariant();
         }
+
+        /// <summary>
+        /// Checks the store directory has the correct permissions.
+        /// </summary>
+        /// <param name="path">
+        /// The path of the directory to check.
+        /// </param>
+        /// <param name="userId">
+        /// The current userId from GetEUid().
+        /// </param>
+        private static void EnsureDirectoryPermissions(string path, uint userId)
+        {
+            int mode = GetFileMode(path, userId, isDirectory: true);
+            if ((mode & (int)Interop.Sys.Permissions.Mask) != (int)Interop.Sys.Permissions.S_IRWXU)
+            {
+                throw new CryptographicException(SR.Format(SR.Cryptography_InvalidDirectoryPermissions, path));
+            }
+        }
+
+        /// <summary>
+        /// Checks the file has the correct permissions.
+        /// </summary>
+        /// <param name="path">
+        /// The path of the file to check.
+        /// </param>
+        /// <param name="userId">
+        /// The current userId from GetEUid().
+        /// </param>
+        private static void EnsureFilePermissions(string path, uint userId)
+        {
+            // Verify that we're creating files with u+rw and o-rw, g-rw.
+            const Interop.Sys.Permissions requiredPermissions =
+                Interop.Sys.Permissions.S_IRUSR |
+                Interop.Sys.Permissions.S_IWUSR;
+
+            const Interop.Sys.Permissions forbiddenPermissions =
+                Interop.Sys.Permissions.S_IROTH |
+                Interop.Sys.Permissions.S_IWOTH |
+                Interop.Sys.Permissions.S_IRGRP |
+                Interop.Sys.Permissions.S_IWGRP;
+
+            int mode = GetFileMode(path, userId, isDirectory: false);
+
+            if ((mode & (int)requiredPermissions) != (int)requiredPermissions)
+            {
+                throw new CryptographicException(SR.Format(SR.Cryptography_InsufficientFilePermissions, path));
+            }
+
+            if ((mode & (int)forbiddenPermissions) != 0)
+            {
+                throw new CryptographicException(SR.Format(SR.Cryptography_TooBroadFilePermissions, path));
+            }
+        }
+
+        private static int GetFileMode(string path, uint userId, bool isDirectory)
+        {
+            Interop.Sys.FileStatus stat;
+            if (Interop.Sys.Stat(path, out stat) != 0)
+            {
+                throw new CryptographicException(
+                    SR.Cryptography_FileStatusError,
+                    Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), path, isDirectory));
```

Comment:
I'm still not sure it's worth it.  Since it's being wrapped in a CryptographicException, I might suggest just doing:
```C#
Interop.Sys.ErrorInfo error = Interop.Sys.GetLastErrorInfo();
throw new CryptographicException(
    SR.Cryptography_FileStatusError,
    new IOException(error.GetErrorMessage(), error.RawErrno));
```